File: projects/mcp-8004-enterprise/agents_implementation/aml-bridge.js
/**
 * aml-bridge.js  [BANK DMZ EXTERNAL — bidirectional]
 *
 * Watches AMLOracle events and routes them to/from the aml-server MCP server.
 *
 * Event flows:
 *   AMLReviewRequested  → callMcpTool('screen_client')
 *     → action='request_documents' → requestClientData()
 *     → action='submit_recommendation' → submitRecommendation()
 *
 *   DataFulfilled       → callMcpTool('continue_screening')
 *     → action='request_documents' → requestClientData() (next round)
 *     → action='submit_recommendation' → submitRecommendation()
 *
 * Usage:
 *   node aml-bridge.js \
 *     --contract  0xAMLOracleAddress \
 *     --rpc       http://127.0.0.1:8545 \
 *     --privkey   0xBankAMLAgentKey \
 *     --agent-id  0
 *
 * ENV equivalents: AML_CONTRACT_ADDRESS, RPC_URL, ORACLE_PRIVATE_KEY, AML_AGENT_ID
 * Optional governance: --flow-auth, --reputation-gate, --autonomy-bounds, --action-permit, --identity-registry
 */

import { ethers } from 'ethers';
import { bootstrapBridge, callMcpTool, governancePreflight, arg } from './bridge-base.js';

const LABEL          = 'aml-bridge';
const MCP_ENDPOINT   = arg('--mcp-endpoint', 'AML_MCP_ENDPOINT') ?? 'http://localhost:8010';
const CAP_AML_REVIEW = ethers.id('aml_review');
const TOOL_SCREEN    = ethers.id('screen_client');

const ABI = [
  'event AMLReviewRequested(bytes32 indexed requestId, bytes32 indexed flowId, uint256 bankAgentId, uint256 clientAgentId, uint256 timestamp)',
  'event DataFulfilled(bytes32 indexed requestId, bytes32 indexed flowId, bytes32 dataHash, uint256 submittingAgentId, uint256 timestamp)',
  'function requestClientData(bytes32 requestId, uint256 bankAgentId, bytes32 dataSpecHash)',
  'function submitRecommendation(bytes32 requestId, uint256 bankAgentId, bytes32 resultHash)',
];

async function handleReviewRequested(oracle, agentId, contracts, event) {
  const { requestId, flowId, clientAgentId } = event.args;
  const traceId = flowId; // flowId IS the traceId in this architecture

  console.log(`\n[${LABEL}] ← AMLReviewRequested  requestId=${requestId}  flowId=${flowId}`);

  const ok = await governancePreflight(LABEL, { flowId, agentId, capability: CAP_AML_REVIEW, toolHash: TOOL_SCREEN, contracts });
  if (!ok) return;

  try {
    const result = await callMcpTool(MCP_ENDPOINT, 'screen_client', {
      flow_id:         flowId,
      request_id:      requestId,
      client_agent_id: clientAgentId.toString(),
    }, traceId);

    if (result.action === 'request_documents') {
      console.log(`[${LABEL}]   → requestClientData  spec=${result.spec_hash}`);
      const tx = await oracle.requestClientData(requestId, agentId, result.spec_hash);
      await tx.wait();
      console.log(`[${LABEL}]   ✓ requestClientData  tx=${tx.hash}`);
    } else {
      console.log(`[${LABEL}]   → submitRecommendation  cleared=${result.cleared}`);
      const tx = await oracle.submitRecommendation(requestId, agentId, result.result_hash);
      await tx.wait();
      console.log(`[${LABEL}]   ✓ submitRecommendation  tx=${tx.hash}`);
    }
  } catch (err) {
    console.error(`[${LABEL}]   ✗ ${err.message}`);
  }
}

async function handleDataFulfilled(oracle, agentId, contracts, event) {
  const { requestId, flowId, dataHash, submittingAgentId } = event.args;
  const traceId = flowId;

  // Only respond to data submitted by the client agent (not our own fulfillments)
  if (submittingAgentId === agentId) return;

  console.log(`\n[${LABEL}] ← DataFulfilled  requestId=${requestId}  dataHash=${dataHash}`);

  const ok = await governancePreflight(LABEL, { flowId, agentId, capability: CAP_AML_REVIEW, toolHash: TOOL_SCREEN, contracts });
  if (!ok) return;

  try {
    // We need the current round; read from oracle state or pass round via event.
    // For simplicity, pass round=1 and let the server handle it.
    const result = await callMcpTool(MCP_ENDPOINT, 'continue_screening', {
      flow_id:    flowId,
      request_id: requestId,
      data_hash:  dataHash,
      round:      1,
    }, traceId);

    if (result.action === 'request_documents') {
      console.log(`[${LABEL}]   → requestClientData (round 2)  spec=${result.spec_hash}`);
      const tx = await oracle.requestClientData(requestId, agentId, result.spec_hash);
      await tx.wait();
      console.log(`[${LABEL}]   ✓ requestClientData  tx=${tx.hash}`);
    } else {
      console.log(`[${LABEL}]   → submitRecommendation  cleared=${result.cleared}`);
      const tx = await oracle.submitRecommendation(requestId, agentId, result.result_hash);
      await tx.wait();
      console.log(`[${LABEL}]   ✓ submitRecommendation  tx=${tx.hash}`);
    }
  } catch (err) {
    console.error(`[${LABEL}]   ✗ ${err.message}`);
  }
}

async function main() {
  const { wallet, contracts, agentId, CONTRACT_ADDRESS } = await bootstrapBridge(LABEL, {
    contractFlag: '--contract', contractEnv: 'AML_CONTRACT_ADDRESS',
    agentIdEnv: 'AML_AGENT_ID',
    cardGlob: f => f === 'bank-aml-agent.json',
  });

  const oracle = new ethers.Contract(CONTRACT_ADDRESS, ABI, wallet);

  console.log(`[${LABEL}] MCP endpoint : ${MCP_ENDPOINT}`);
  console.log(`[${LABEL}] Listening for AMLOracle events…\n`);

  oracle.on('AMLReviewRequested', (...args) => {
    const event = args[args.length - 1];
    handleReviewRequested(oracle, agentId, contracts, event);
  });

  oracle.on('DataFulfilled', (...args) => {
    const event = args[args.length - 1];
    handleDataFulfilled(oracle, agentId, contracts, event);
  });
}

main().catch(err => { console.error(err); process.exit(1); });

File: examples/mcp-erc8004-enterprise/agents_implementation_py/bridges/aml_bridge.py
"""
aml_bridge.py — Port of aml-bridge.js using web3.py AsyncWeb3.

Watches AMLOracle events and routes to/from the Python AML MCP server.

Event flows:
  AMLReviewRequested → screen_client
    → request_documents → requestClientData()
    → submit_recommendation → submitRecommendation()
  DataFulfilled → continue_screening
    → same actions

Usage:
    python -m bridges.aml_bridge \
        --contract 0x... --rpc http://... --privkey 0x... --agent-id 0
ENV: AML_CONTRACT_ADDRESS, RPC_URL, ORACLE_PRIVATE_KEY, AML_AGENT_ID
"""

from __future__ import annotations

import asyncio
import logging
import os

from web3 import Web3

from shared.abis import AML_ORACLE_ABI
from shared.bridge_base import arg, bootstrap_bridge, call_mcp_tool, governance_preflight

logger = logging.getLogger(__name__)

LABEL        = "aml-bridge"
MCP_ENDPOINT = arg("--mcp-endpoint", "AML_MCP_ENDPOINT", "http://localhost:8010")

CAP_AML_REVIEW = Web3.keccak(text="aml_review")
TOOL_SCREEN    = Web3.keccak(text="screen_client")


async def _handle_review_requested(oracle, agent_id: int, contracts: dict, event) -> None:
    args = event["args"]
    request_id    = args["requestId"]
    flow_id       = args["flowId"]
    client_agent_id = args["clientAgentId"]
    trace_id      = flow_id.hex() if isinstance(flow_id, bytes) else flow_id

    logger.info("\n[%s] ← AMLReviewRequested  requestId=%s  flowId=%s", LABEL, request_id.hex(), trace_id)

    ok = await governance_preflight(
        LABEL,
        flow_id=flow_id if isinstance(flow_id, bytes) else bytes.fromhex(flow_id.removeprefix("0x")),
        agent_id=agent_id,
        capability=CAP_AML_REVIEW,
        tool_hash=TOOL_SCREEN,
        contracts=contracts,
    )
    if not ok:
        return

    try:
        result = await call_mcp_tool(
            MCP_ENDPOINT,
            "screen_client",
            {
                "flow_id":         trace_id,
                "request_id":      request_id.hex(),
                "client_agent_id": str(client_agent_id),
            },
            trace_id,
        )

        req_id_bytes = request_id if isinstance(request_id, bytes) else bytes.fromhex(request_id.removeprefix("0x"))
        if result["action"] == "request_documents":
            logger.info("[%s]   → requestClientData  spec=%s", LABEL, result["spec_hash"])
            spec = bytes.fromhex(result["spec_hash"].removeprefix("0x"))
            tx_hash = await oracle.functions.requestClientData(req_id_bytes, agent_id, spec).transact()
            logger.info("[%s]   ✓ requestClientData  tx=%s", LABEL, tx_hash.hex())
        else:
            logger.info("[%s]   → submitRecommendation  cleared=%s", LABEL, result.get("cleared"))
            res_hash = bytes.fromhex(result["result_hash"].removeprefix("0x"))
            tx_hash = await oracle.functions.submitRecommendation(req_id_bytes, agent_id, res_hash).transact()
            logger.info("[%s]   ✓ submitRecommendation  tx=%s", LABEL, tx_hash.hex())
    except Exception as exc:
        logger.error("[%s]   ✗ %s", LABEL, exc)


async def _handle_data_fulfilled(oracle, agent_id: int, contracts: dict, event) -> None:
    args = event["args"]
    request_id       = args["requestId"]
    flow_id          = args["flowId"]
    data_hash        = args["dataHash"]
    submitting_agent = args["submittingAgentId"]
    trace_id         = flow_id.hex() if isinstance(flow_id, bytes) else flow_id

    if submitting_agent == agent_id:
        return  # our own fulfillment

    logger.info("\n[%s] ← DataFulfilled  requestId=%s  dataHash=%s", LABEL, request_id.hex(), data_hash.hex())

    ok = await governance_preflight(
        LABEL,
        flow_id=flow_id if isinstance(flow_id, bytes) else bytes.fromhex(flow_id.removeprefix("0x")),
        agent_id=agent_id,
        capability=CAP_AML_REVIEW,
        tool_hash=TOOL_SCREEN,
        contracts=contracts,
    )
    if not ok:
        return

    try:
        result = await call_mcp_tool(
            MCP_ENDPOINT,
            "continue_screening",
            {
                "flow_id":    trace_id,
                "request_id": request_id.hex(),
                "data_hash":  data_hash.hex(),
                "round":      1,
            },
            trace_id,
        )

        req_id_bytes = request_id if isinstance(request_id, bytes) else bytes.fromhex(request_id.removeprefix("0x"))
        if result["action"] == "request_documents":
            spec = bytes.fromhex(result["spec_hash"].removeprefix("0x"))
            tx_hash = await oracle.functions.requestClientData(req_id_bytes, agent_id, spec).transact()
            logger.info("[%s]   ✓ requestClientData (round 2)  tx=%s", LABEL, tx_hash.hex())
        else:
            res_hash = bytes.fromhex(result["result_hash"].removeprefix("0x"))
            tx_hash = await oracle.functions.submitRecommendation(req_id_bytes, agent_id, res_hash).transact()
            logger.info("[%s]   ✓ submitRecommendation  tx=%s", LABEL, tx_hash.hex())
    except Exception as exc:
        logger.error("[%s]   ✗ %s", LABEL, exc)


async def main() -> None:
    ctx = await bootstrap_bridge(
        LABEL,
        contract_flag="--contract",
        contract_env="AML_CONTRACT_ADDRESS",
        agent_id_env="AML_AGENT_ID",
        card_glob="bank-aml-agent.json",
        prompt_name="bank-aml-agent",
        prompt_version_env="AML_PROMPT_VERSION",
    )

    oracle = ctx.w3.eth.contract(address=ctx.contract_address, abi=AML_ORACLE_ABI)

    logger.info("[%s] MCP endpoint : %s", LABEL, MCP_ENDPOINT)
    logger.info("[%s] Listening for AMLOracle events…", LABEL)

    # Poll for events (web3.py async event polling)
    review_filter = await oracle.events.AMLReviewRequested.create_filter(from_block="latest")
    data_filter   = await oracle.events.DataFulfilled.create_filter(from_block="latest")

    while True:
        for event in await review_filter.get_new_entries():
            asyncio.create_task(
                _handle_review_requested(oracle, ctx.agent_id, ctx.contracts, event)
            )
        for event in await data_filter.get_new_entries():
            asyncio.create_task(
                _handle_data_fulfilled(oracle, ctx.agent_id, ctx.contracts, event)
            )
        await asyncio.sleep(2)


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(main())
